addons/test_mail/tests/test_mail_activity.py

```diff
@@ -333,3 +349,177 @@ def test_activity_mixin_reschedule_user(self):
             user_id=self.user_admin.id,
             new_user_id=self.user_employee.id)
         self.assertEqual(rec.activity_ids[0].user_id, self.user_employee)
+
+    def test_mail_activity_state(self):
+        """Create 3 activity for 2 different users in 2 different timezones.
+
+        User UTC (+0h)
+        User Australia (+11h)
+        Today datetime: 1/1/2020 16h
+
+        Activity 1 & User UTC
+            1/1/2020 - 16h UTC       -> The state is today
+
+        Activity 2 & User Australia
+            1/1/2020 - 16h UTC
+            2/1/2020 -  1h Australia -> State is overdue
+
+        Activity 3 & User UTC
+            1/1/2020 - 23h UTC       -> The state is today
+        """
+        today_utc = datetime(2020, 1, 1, 16, 0, 0)
+
+        class MockedDatetime(datetime):
+            @classmethod
+            def utcnow(cls):
+                return today_utc
+
+        record = self.env['mail.test.activity'].create({'name': 'Record'})
+
+        with patch('odoo.addons.mail.models.mail_activity.datetime', MockedDatetime):
+            activity_1 = self.env['mail.activity'].create({
+                'summary': 'Test',
+                'activity_type_id': 1,
+                'res_model_id': self.env.ref('test_mail.model_mail_test_activity').id,
+                'res_id': record.id,
+                'date_deadline': today_utc,
+                'user_id': self.user_utc.id,
+            })
+
+            activity_2 = activity_1.copy()
+            activity_2.user_id = self.user_australia
+            activity_3 = activity_1.copy()
+            activity_3.date_deadline += relativedelta(hours=7)
+
+            self.assertEqual(activity_1.state, 'today')
+            self.assertEqual(activity_2.state, 'overdue')
+            self.assertEqual(activity_3.state, 'today')
+
+    def test_mail_activity_mixin_search_state_basic(self):
+        """Test the search method on the "activity_state".
+
+        Test all the operators and also test the case where the "activity_state" is
+        different because of the timezone. There's also a tricky case for which we
+        "reverse" the domain for performance purpose.
+        """
+        today_utc = datetime(2020, 1, 1, 16, 0, 0)
+
+        class MockedDatetime(datetime):
+            @classmethod
+            def utcnow(cls):
+                return today_utc
+
+        # Create some records without activity schedule on it for testing
+        self.env['mail.test.activity'].create([
```

Comment:
Well we could maybe create less records, if we play with 2 records for state, maybe 2 or 3 additional one are sufficient.